lib/ably/realtime/channel.rb

```diff
@@ -179,15 +179,28 @@ def presence
         @presence
       end
 
-      # Return the message history of the channel
+      # Return the message history of the channel.
+      #
+      # Once attached to a channel, you can retrieve messages published on the channel before the
+      # channel was attached with the option <tt>:end => :before_attach</tt>.  This is very useful for
+      # developers who wish to display both new real-time messages as well as historical messages with
+      # the guarantee that no messages have been missed.
+      #
       #
       # @param (see Ably::Rest::Channel#history)
       # @option options (see Ably::Rest::Channel#history)
+      # @option options [Integer,Time,Symbol]   :end        Time or millisecond since epoch, or +:before_attach+ for all messages published before this channel has been attached.
       #
       # @yield [Ably::Models::PaginatedResource<Ably::Models::Message>] An Array of {Ably::Models::Message} objects that supports paging (#next_page, #first_page)
       #
       # @return [Ably::Util::SafeDeferrable]
       def history(options = {}, &callback)
+        if options[:end] == :before_attach
+          raise ArgumentError, 'option :before_attach cannot be specified if the channel is not attached' unless attached?
```

Comment:
I'm sorry but that's not true. Your async wrapper does a websocket connect, and `subscribe`/`publish` are doing an implicit `attach`. I don't think I was far off thinking that `history` required the same.

I would have to compare your `subscribe`/`publish` implementation, and see what is inside of the state machine to know what is going on. So no. It's not that simple.